Function type consistency between goto programs and symbol table

xbauch opened the pull request:
A simple iteration over goto-functions to check that it's base type is the same
as the one stored in the symbol table.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe committed:
Add macros for goto model validation

Commit:
Function type consistency between goto programs and symbol table

A simple iteration over goto-functions to check that it's base type is the same
as the one stored in the symbol table.

smowton commented on src/goto-programs/goto_model.h:
```diff
@@ -95,6 +97,27 @@ class goto_modelt : public abstract_goto_modelt
+      if(!base_type_eq(
+           it->second.type, symbol_table.lookup_ref(it->first).type, ns))
+      {
+        msg.error() << "error" << messaget::eom;
```
Say what's wrong!

NlightNFotis replied:
Yes, a better error message here would be good.

smowton commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -515,6 +515,10 @@ int cbmc_parse_optionst::doit()
     return get_goto_program_ret;
 
+  INVARIANT(
+    !goto_model.check_internal_invariants(*this),
```
`check_consistency`?

NlightNFotis replied:
I find these a little bit confusing in design, both here and in #3118. Do you think it would be better if `check_internal_invariants` returned `true` by default, as the *all is good* case, and `false` otherwise? This invariant would also look cleaner as well, as it would just read as `the internal invariants of the goto model should always hold, which this is what the invariant will check by essence of it returning true` and not `if this doesn't return false, then the internal invariants hold`. Like, it's the double negative both in the condition here, and the return values of `check_internal_invariants` that's slightly confusing, as it's inverse logic to what's expected. But that's just me, certainly not required.

xbauch replied:
@smowton Within this PR, yes. But we are preparing a larger collection of checks and not all of them target consistency.

xbauch replied:
@NlightNFotis it was my understanding that we prefer returning `true` when things fail. It can of course be negated.

owen-mc-diffblue replied:
@NlightNFotis My understanding is that linux programs return 0 to indicate no error and a non-zero integer to represent some kind of error. C programs copy this and return 0 to indicate no error. We follow this by returning `false` to indicate no error and `true` to indicate error. I find it confusing too.

danpoe replied:
Yeah most functions return `true` on error and `false` on success. Let's stick with this convention for consistency for now.

NlightNFotis reviewed (commented):
A couple of comments.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 8558e891962d1603ec9a82633cee1d21f077213b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87310747

Commit:
Better error message

Now reports the id of violating function and the two inconsistent types.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 150cf46d1a04b50d066293b5224e109ef072190e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87323737

danpoe commented on src/goto-programs/goto_model.h:
```diff
@@ -95,6 +97,33 @@ class goto_modelt : public abstract_goto_modelt
+    namespacet ns(symbol_table);
+    forall_goto_functions(it, goto_functions)
+    {
+      if(!base_type_eq(
```
Could use the `DATA_CHECK` macros.

Commit:
Reworked based on PR #3123

Using DATA_CHECK instead of calling error handling directly.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 43052d570a1c86daaf5084e22ed34f71df3b4b8c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87497802

Commit:
Added unit test for wrong type in table

Builds a goto model with one function and checks that: 1) with the right type
in the table the validation succeeds, 2) with the wrong type the validation
fails.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 00357ff365bccb68cf4395653ead5b2f95d50c13).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87657183

hannes-steffenhagen-diffblue commented on src/util/validate.h:
```diff
@@ -0,0 +1,61 @@
+/// goto programs, expressions, instructions, etc. Based on the value of the
+/// variable vm (the validation mode), it either uses DATA_INVARIANT() to check
+/// those conditions, or throws an exception when a condition does not hold.
+#define DATA_CHECK(condition, message)                                         \
```
I'm unclear why we want this. Either something is an invariant or it isn't, I don't see the value in having a single construct for both cases.

xbauch commented:
Moved to #3118.
